T7353: T7360: netplug: behavior change 1.3.8 -> 1.4 when interface with DHCP address looses carrier

c-po opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->

There is a behavior change from VyOS 1.3.8 -> 1.4.0 as when an interface configured with DHCP will loose it's carrier in VyOS 1.3.8 dhclient is stopped and the IP address is removed from the interface.

In VyOS 1.4 the DHCP assigned IP address stays active on the interface that just lost it's carrier.

In addition this fixes a bug that was present in VyOS 1.3.8 where the DHCP assigned IP address only got cleared on base interfaces like eth0 but not on any VLAN interface e.g. eth0.10.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
* https://vyos.dev/T7353
* https://vyos.dev/T7360

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->
* https://github.com/vyos/vyos-build/pull/944

## How to test / Smoketest result

### DHCP server on VyOS 1.4.2

```
set interfaces ethernet eth1 vif 1111 address '192.0.2.1/24'
set interfaces ethernet eth1 vif 1111 address '2001:db8:1111::1/64'

set service dhcp-server shared-network-name TEST subnet 192.0.2.0/24 default-router '192.0.2.1'
set service dhcp-server shared-network-name TEST subnet 192.0.2.0/24 range 1 start '192.0.2.100'
set service dhcp-server shared-network-name TEST subnet 192.0.2.0/24 range 1 stop '192.0.2.200'
set service dhcpv6-server shared-network-name TEST subnet 2001:db8:1111::/64 address-range start 2001:db8:1111::f000 stop '2001:db8:1111::ffff'
```

### DUT

```
set interfaces ethernet eth1 vif 1111 address 'dhcp'
set interfaces ethernet eth1 vif 1111 address 'dhcpv6'
```

will give

```
vyos@vyos:~$ show interfaces
Codes: S - State, L - Link, u - Up, D - Down, A - Admin Down
Interface    IP Address               MAC                VRF        MTU  S/L    Description
-----------  -----------------------  -----------------  -------  -----  -----  -------------
eth1         -                        00:50:56:b3:38:c5  default   1500  u/u
eth1.1111    192.0.2.100/24           00:50:56:b3:38:c5  default   1500  u/u
             2001:db8:1111::fde4/128
```

#### "pulling" the plug on eth1

```
Apr 14 21:22:00 netplugd[2516]: eth1: state ACTIVE flags 0x00011043 UP,BROADCAST,RUNNING,MULTICAST,10000 -> 0x00001003 UP,BROADCAST,MULTICAST
Apr 14 21:22:00 netplugd[3104]: /etc/netplug/netplug eth1 out -> pid 3104
Apr 14 21:22:00 netplugd[2516]: eth1.1111: state ACTIVE flags 0x00011043 UP,BROADCAST,RUNNING,MULTICAST,10000 -> 0x00001003 UP,BROADCAST,MULTICAST
Apr 14 21:22:00 netplugd[3105]: /etc/netplug/netplug eth1.1111 out -> pid 3105
Apr 14 21:22:00 systemd[1]: Stopping dhclient@eth1.1111.service - DHCP client on eth1.1111...
Apr 14 21:22:01 systemd[1]: dhclient@eth1.1111.service: Deactivated successfully.
Apr 14 21:22:01 systemd[1]: Stopped dhclient@eth1.1111.service - DHCP client on eth1.1111.
Apr 14 21:22:01 dhcp6c[2651]: release_all_ia: Start address release
Apr 14 21:22:01 dhcp6c[2651]: release_ia: release an IA: NA-0
Apr 14 21:22:01 systemd[1]: Stopping dhcp6c@eth1.1111.service - WIDE DHCPv6 client on eth1.1111...
Apr 14 21:22:30 systemd[1]: dhcp6c@eth1.1111.service: Deactivated successfully.
Apr 14 21:22:30 systemd[1]: Stopped dhcp6c@eth1.1111.service - WIDE DHCPv6 client on eth1.1111.
```

```
vyos@vyos:~$ show interfaces
Codes: S - State, L - Link, u - Up, D - Down, A - Admin Down
Interface    IP Address         MAC                VRF        MTU  S/L    Description
-----------  -----------------  -----------------  -------  -----  -----  -------------
eth1         -                  00:50:56:b3:38:c5  default   1500  u/D
eth1.1111    -                  00:50:56:b3:38:c5  default   1500  u/D
```

#### "inserting" the plug on eth1

```
Apr 14 21:24:14 netplugd[2516]: eth1: state INACTIVE flags 0x00001003 UP,BROADCAST,MULTICAST -> 0x00011043 UP,BROADCAST,RUNNING,MULTICAST,10000
Apr 14 21:24:14 netplugd[2516]: eth1.1111: state INACTIVE flags 0x00001003 UP,BROADCAST,MULTICAST -> 0x00011043 UP,BROADCAST,RUNNING,MULTICAST,10000
Apr 14 21:24:15 systemd[1]: Starting dhclient@eth1.1111.service - DHCP client on eth1.1111...
Apr 14 21:24:15 systemd[1]: Started dhclient@eth1.1111.service - DHCP client on eth1.1111.
Apr 14 21:24:15 systemd[1]: Starting dhcp6c@eth1.1111.service - WIDE DHCPv6 client on eth1.1111...
Apr 14 21:24:15 systemd[1]: Started dhcp6c@eth1.1111.service - WIDE DHCPv6 client on eth1.1111.
```

```
vyos@vyos:~$ show interfaces
Codes: S - State, L - Link, u - Up, D - Down, A - Admin Down
Interface    IP Address               MAC                VRF        MTU  S/L    Description
-----------  -----------------------  -----------------  -------  -----  -----  -------------
eth1         -                        00:50:56:b3:38:c5  default   1500  u/u
eth1.1111    192.0.2.100/24           00:50:56:b3:38:c5  default   1500  u/u
             2001:db8:1111::fde4/128
```


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [ ] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

c-po committed:
T7353: honor service startup order for vyos-grub-update

vyos-grub-update service must have been started prior reaching systemd
vyos.target.

c-po committed:
netplug: T7353: define proper systemd service

Netplug daemon is started last after all interfaces got initialized to
properly monitor them for up/down events

c-po committed:
netplug: T7353: bugfix DHCP client not stopped when interface goes down

The initial rewrite of vyos-netplug-dhcp-client by me tried to outsmart the
DHCP client implementation by re-using vyos.ifconfig.Interface(). This added
a regression where an interface loosing it's carrier no longer deconfigured
it's IP address. This was a change in behavior form VyOS 1.3.

In addition a bug is fixed as when a VLANs interface parent looses it's carrier
we now also stop the DHCP client process.

This script is now back to simply starting/stopping the DHCP client process.

sarthurdev reviewed (approved):
Adds netplug systemd unit, starts service after `vyos-router`.

c-po committed:
netplug: T7360: DHCPv6 address is not cleared when interface goes oper-down

When an interface goes down - e.g. cable unplugged - the DHCPv6 assigned IPv6
address is not removed from said interface. We should provide the same behavior
as with IPv4. IPv6 address should be removed and dhcpv6 client restarted once
the interface goes operational up again.

dmbaturin reviewed (approved):
The logic looks correct, the fixes are needed, and I trust the manual testing report.
